src/pages/api/discord.ts

```diff
@@ -0,0 +1,214 @@
+}
+
+async function fetchEggheadUser(token: string) {
+  const current: Viewer = await got
+    .get(`${EGGHEAD_AUTH_DOMAIN}/api/v1/users/current`, {
+      responseType: 'json',
+      headers: {
+        Authorization: `Bearer ${token}`,
+      },
+      hooks: {beforeError: [handleGotError]},
+    })
+    .json()
+
+  return current
+}
+
+type UpdateDiscordUserForEggheadParams = {
+  discordMember: GuildMember
+  eggheadUser: Viewer
+  eggheadToken: string
+}
+
+async function updateDiscordRolesForEggheadUser({
+  discordMember,
+  eggheadUser,
+  eggheadToken,
+}: UpdateDiscordUserForEggheadParams) {
+  if (discordMember && eggheadUser) {
+    const updateEgghead: Viewer = await got
+      .put(`${EGGHEAD_AUTH_DOMAIN}/api/v1/users/${eggheadUser.id}`, {
+        responseType: 'json',
+        json: {
+          user: {discord_id: discordMember.id},
+        },
+        headers: {
+          Authorization: `Bearer ${eggheadToken}`,
+        },
+        hooks: {beforeError: [handleGotError]},
+      })
+      .json()
+
+    if (updateEgghead.is_pro && DISCORD_MEMBER_ROLE_ID) {
+      await discordMember.roles.add(DISCORD_MEMBER_ROLE_ID)
+    } else if (DISCORD_MEMBER_ROLE_ID) {
+      await discordMember.roles.remove(DISCORD_MEMBER_ROLE_ID)
+    }
+
+    return updateEgghead
+  } else {
+    throw new Error(`requires egghead and discord users to ipdate`)
+  }
+}
+
+async function getDiscordBotClient() {
+  const client = new Client()
+  await client.login(DISCORD_BOT_TOKEN)
+  return client
+}
+
+async function fetchOrInviteDiscordMember(
+  client: Client,
+  discordUser: User,
+  discordToken: any,
+): Promise<GuildMember> {
+  const guild: Guild | undefined = client.guilds.cache.get(
+    DISCORD_GUILD_ID as string,
+  )
+  if (guild) {
+    const discordMember: GuildMember = await guild.members
+      .fetch(discordUser.id)
+      .catch(async (e) => {
+        await inviteToServer({
+          discordUserId: discordUser.id,
+          discordUserToken: discordToken.access_token,
+        })
+
+        const newMember = await fetchOrInviteDiscordMember(
+          client,
+          discordUser,
+          discordToken,
+        )
+        return newMember
+      })
+
+    return discordMember
+  } else {
+    throw new Error('no discord guild was found')
+  }
+}
+
+export default async (req: NextApiRequest, res: NextApiResponse) => {
+  if (req.method === 'POST') {
+    try {
+      const client = await getDiscordBotClient()
+      const {discordUser, discordToken} = await loadDiscordUser(req.body.code)
+      const {eggheadToken} = getTokenFromCookieHeaders(
+        req.headers.cookie as string,
+      )
+
+      if (discordUser && eggheadToken) {
+        const discordMember: GuildMember = await fetchOrInviteDiscordMember(
+          client,
+          discordUser,
+          discordToken,
+        )
+
+        const eggheadUser: Viewer = await fetchEggheadUser(eggheadToken)
+
+        await updateDiscordRolesForEggheadUser({
```

Comment:
now we will check the egghead user status and add a `member` role if they qualify